File: android/ilds-design-system/src/main/kotlin/com/icicilombard/ilds/components/IldsChip.kt
package com.icicilombard.ilds.components

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Close
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.LocalContentColor
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.Immutable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.selected
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.icicilombard.ilds.tokens.IldsTokens

// Figma "Tag" filter chip — component set 14018:6786; mirrors lib/ilds_chip.dart + ios/IldsChip.swift.

enum class IldsChipSize { Large, Medium }

@Composable
fun IldsChip(
    label: String,
    modifier: Modifier = Modifier,
    size: IldsChipSize = IldsChipSize.Large,
    isSelected: Boolean = false,
    isDisabled: Boolean = false,
    hasPrefixIcon: Boolean = false,
    prefixIcon: (@Composable () -> Unit)? = null,
    hasSuffixButton: Boolean = false,
    onPress: (() -> Unit)? = null,
    onRemove: (() -> Unit)? = null,
) {
    val colors = IldsChipColors.resolve(isSelected, isDisabled)
    val metrics = IldsChipMetrics(size)
    val shape = RoundedCornerShape(IldsTokens.radiusMedium)
    val interactive = !isDisabled && onPress != null && !hasSuffixButton

    Surface(
        modifier = modifier
            .semantics {
                contentDescription = "$label chip"
                if (isSelected) selected = true
                if (interactive) role = Role.Button
            }
            .then(
                if (interactive) {
                    Modifier.clickable(enabled = true, onClick = onPress!!)
                } else {
                    Modifier
                },
            ),
        shape = shape,
        color = colors.background,
        border = BorderStroke(0.5.dp, colors.border),
    ) {
        Row(
            modifier = Modifier
                .height(metrics.height)
                .padding(metrics.padding),
            horizontalArrangement = Arrangement.spacedBy(metrics.gap),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            if (hasPrefixIcon && prefixIcon != null) {
                androidx.compose.runtime.CompositionLocalProvider(
                    LocalContentColor provides colors.label,
                ) {
                    androidx.compose.foundation.layout.Box(
                        modifier = Modifier
                            .size(metrics.iconSlot)
                            .padding(top = 2.dp),
                        contentAlignment = Alignment.Center,
                    ) {
                        prefixIcon()
                    }
                }
            }
            Text(
                text = label,
                style = metrics.textStyle,
                color = colors.label,
                maxLines = 1,
            )
            if (hasSuffixButton) {
                IconButton(
                    onClick = { if (!isDisabled) onRemove?.invoke() },
                    enabled = !isDisabled,
                    modifier = Modifier.size(metrics.iconSlot),
                ) {
                    Icon(
                        imageVector = Icons.Filled.Close,
                        contentDescription = "Remove $label",
                        tint = colors.label,
                        modifier = Modifier.size(metrics.iconSlot - 2.dp),
                    )
                }
            }
        }
    }
}

@Immutable
private data class IldsChipColors(
    val background: Color,
    val border: Color,
    val label: Color,
) {
    companion object {
        fun resolve(isSelected: Boolean, isDisabled: Boolean): IldsChipColors {
            if (isDisabled) {
                return IldsChipColors(
                    IldsTokens.neutralCoolgray200,
                    IldsTokens.neutralCoolgray300,
                    IldsTokens.neutralCoolgray500,
                )
            }
            if (isSelected) {
                return IldsChipColors(
                    IldsTokens.primaryOrange50,
                    IldsTokens.primaryOrange500,
                    IldsTokens.neutralCoolgray900,
                )
            }
            return IldsChipColors(
                IldsTokens.globalWhite000,
                IldsTokens.neutralCoolgray500,
                IldsTokens.neutralCoolgray900,
            )
        }
    }
}

@Immutable
private data class IldsChipMetrics(
    val height: Dp,
    val padding: PaddingValues,
    val gap: Dp,
    val iconSlot: Dp,
    val textStyle: TextStyle,
) {
    constructor(size: IldsChipSize) : this(
        height = when (size) {
            IldsChipSize.Large -> 24.dp
            IldsChipSize.Medium -> 20.dp
        },
        padding = when (size) {
            IldsChipSize.Large -> PaddingValues(horizontal = IldsTokens.sp8, vertical = IldsTokens.sp4)
            IldsChipSize.Medium -> PaddingValues(horizontal = IldsTokens.sp4, vertical = IldsTokens.sp2)
        },
        gap = when (size) {
            IldsChipSize.Large -> IldsTokens.sp4
            IldsChipSize.Medium -> IldsTokens.sp2
        },
        iconSlot = 12.dp,
        textStyle = TextStyle(
            fontSize = IldsTokens.fontSize12,
            lineHeight = 16.sp,
            fontWeight = IldsTokens.fontWeightRegular,
        ),
    )
}

File: ios/Sources/ILDSDesignSystem/IldsChip.swift
import ILDSTokens
import SwiftUI

// Figma "Tag" filter chip — component set 14018:6786; mirrors lib/ilds_chip.dart + web Chip.tsx.

public enum IldsChipSize: Sendable {
    case large, medium
}

public struct IldsChip: View {
    private let label: String
    private let size: IldsChipSize
    private let isSelected: Bool
    private let isDisabled: Bool
    private let hasPrefixIcon: Bool
    private let prefixIcon: AnyView?
    private let hasSuffixButton: Bool
    private let onPress: (() -> Void)?
    private let onRemove: (() -> Void)?

    public init(
        _ label: String,
        size: IldsChipSize = .large,
        isSelected: Bool = false,
        isDisabled: Bool = false,
        hasPrefixIcon: Bool = false,
        prefixIcon: (any View)? = nil,
        hasSuffixButton: Bool = false,
        onPress: (() -> Void)? = nil,
        onRemove: (() -> Void)? = nil
    ) {
        self.label = label
        self.size = size
        self.isSelected = isSelected
        self.isDisabled = isDisabled
        self.hasPrefixIcon = hasPrefixIcon
        self.prefixIcon = prefixIcon.map { AnyView($0) }
        self.hasSuffixButton = hasSuffixButton
        self.onPress = onPress
        self.onRemove = onRemove
    }

    public var body: some View {
        let colors = IldsChipColors.resolve(isSelected: isSelected, isDisabled: isDisabled)
        let metrics = IldsChipMetrics(size: size)

        HStack(spacing: metrics.gap) {
            if hasPrefixIcon, let prefixIcon {
                prefixIcon
                    .foregroundStyle(ILDSTokens.primaryOrange500)
                    .frame(width: metrics.iconSlot, height: metrics.iconSlot)
                    .padding(.top, 2)
            }
            Text(label)
                .font(metrics.font)
                .foregroundStyle(colors.label)
                .lineLimit(1)
            if hasSuffixButton {
                Button(action: { onRemove?() }) {
                    Image(systemName: "xmark.circle")
                        .font(.system(size: metrics.iconSlot))
                        .foregroundStyle(colors.label)
                }
                .buttonStyle(.plain)
                .disabled(isDisabled)
                .padding(.top, 2)
            }
        }
        .frame(height: metrics.height)
        .padding(.horizontal, metrics.horizontalPadding)
        .padding(.vertical, metrics.verticalPadding)
        .background(colors.background)
        .overlay {
            RoundedRectangle(cornerRadius: ILDSTokens.radiusMedium)
                .strokeBorder(colors.border, lineWidth: 0.5)
        }
        .clipShape(RoundedRectangle(cornerRadius: ILDSTokens.radiusMedium))
        .contentShape(RoundedRectangle(cornerRadius: ILDSTokens.radiusMedium))
        .onTapGesture {
            guard !isDisabled, !hasSuffixButton else { return }
            onPress?()
        }
        .accessibilityAddTraits(isSelected ? .isSelected : [])
        .accessibilityLabel("\(label) chip")
        .allowsHitTesting(!isDisabled)
    }
}

private struct IldsChipColors {
    let background: Color
    let border: Color
    let label: Color

    static func resolve(isSelected: Bool, isDisabled: Bool) -> IldsChipColors {
        if isDisabled {
            return IldsChipColors(
                background: ILDSTokens.neutralCoolgray200,
                border: ILDSTokens.neutralCoolgray300,
                label: ILDSTokens.neutralCoolgray500
            )
        }
        if isSelected {
            return IldsChipColors(
                background: ILDSTokens.primaryOrange50,
                border: ILDSTokens.primaryOrange500,
                label: ILDSTokens.neutralCoolgray900
            )
        }
        return IldsChipColors(
            background: ILDSTokens.globalWhite000,
            border: ILDSTokens.neutralCoolgray500,
            label: ILDSTokens.neutralCoolgray900
        )
    }
}

private struct IldsChipMetrics {
    let height: CGFloat
    let horizontalPadding: CGFloat
    let verticalPadding: CGFloat
    let gap: CGFloat
    let iconSlot: CGFloat
    let font: Font

    init(size: IldsChipSize) {
        switch size {
        case .large:
            height = 24
            horizontalPadding = ILDSTokens.sp8
            verticalPadding = ILDSTokens.sp4
            gap = ILDSTokens.sp4
            iconSlot = 12
        case .medium:
            height = 20
            horizontalPadding = ILDSTokens.sp4
            verticalPadding = ILDSTokens.sp2
            gap = ILDSTokens.sp2
            iconSlot = 12
        }
        font = .system(size: ILDSTokens.fontSize12, weight: ILDSTokens.fontWeightRegular)
    }
}
